src/transformers/generation/utils.py

```diff
@@ -1348,52 +1348,68 @@ def _get_initial_cache_position(self, input_ids, model_kwargs):
 
         past_length = 0
         if "past_key_values" in model_kwargs:
-            if isinstance(model_kwargs["past_key_values"], Cache):
-                past_length = model_kwargs["past_key_values"].get_seq_length()
+            past_key_values = model_kwargs["past_key_values"]
+            if self.config.is_encoder_decoder and isinstance(past_key_values[0], Cache):
+                past_key_values = past_key_values[0]
```

Comment:
Looking at the code, `past_key_values[0]` is not very informative. If I was not familiar with the code, I would be wondering "is `[0]` indexing a layer? self/cross attention? if it's self/cross attention, which index is each attention type?"

WDYT of holding the caches in some class that allows naming (`typing.NamedTuple`, `dict`, `dataclass`, ...)? If there are no technical barriers associated with the change, I believe the code would be more readable, i.e. something like `past_key_values.self_attention` or `past_key_values.decoder` instead of `past_key_values[0]` 🤗

Reply:
Agree that indexing 0 is not very informative and `NamedTuple` is a good option. But I was also thinking about other language models that can be used in enc-dec setting and also as a standalone decoder. 

I already tried to make one of them support cache class, imo tuple of cache classes is a bit messy in this case as we have to always check if it's used in enc-dec setting and then try to slice the first cache object. I wanted to propose to add `is_encoder_decoder` as arg when init any cache class so that all decisions about which cache to use can be toggled and maintained from one location. WDYT? Haven't tried it yet, so maybe I'm missing something

Reply:
I see. if/else clauses are definitely annoying. On the other hand, we should avoid adding more functionality to the existing cache classes, to avoid making them complex.

We could build a cache class wrapper for encoder-decoder models, i.e. a class that would live in `cache_utils.py` whose purpose would be:
1. to provide a unified API to this case, by holding the two caches in named attributes (just like a named tuple)
2. mapping common utility calls like `get_seq_length()` to the decoder cache

This would both improve readability (1) and prevent most if/elses that we see here (2). WDYT? (@zucchini-nlp @sanchit-gandhi )

Reply:
Yes, I like this idea!

Reply:
The advantage of an enc-dec wrapper is that it would also facilitate the case where we want a different cache object for the encoder and decoder (e.g. dynamic cache for encoder, static cache for decoder). Adding `is_encoder_decoder` to all the cache classes restricts us to using the same cache for each.

Interestingly, I added such a class in commit https://github.com/huggingface/transformers/commit/93c97c1feff239fee19e8012294d2622a6e3339f, but reverted to the original `tuple` design for simplicity. I'll re-revert this given we're in agreement a wrapper is the best long-term design!